editor/animation_track_editor.cpp

```diff
@@ -3383,74 +3369,82 @@ void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
-				}
-			}
+		if (insert_data[i].type != Animation::TYPE_VALUE) {
+			continue;
 		}
 
-		if (bool(EDITOR_DEF("editors/animation/confirm_insert_track", true))) {
-			//potential new key, does not exist
-			if (num_tracks == 1) {
-				// TRANSLATORS: %s will be replaced by a phrase describing the target of track.
-				insert_confirm_text->set_text(vformat(TTR("Create new track for %s and insert key?"), p_id.query));
-			} else {
-				insert_confirm_text->set_text(vformat(TTR("Create %d new tracks and insert keys?"), num_tracks));
+		switch (insert_data[i].value.get_type()) {
+			case Variant::INT:
+			case Variant::FLOAT:
+			case Variant::VECTOR2:
+			case Variant::VECTOR3:
+			case Variant::QUATERNION:
+			case Variant::PLANE:
+			case Variant::COLOR: {
+				// Valid.
+			} break;
+			default: {
+				all_bezier = false;
 			}
+		}
+	}
 
-			insert_confirm_bezier->set_visible(all_bezier);
-			insert_confirm_reset->set_visible(reset_allowed);
-
-			insert_confirm->get_ok_button()->set_text(TTR("Create"));
-			insert_confirm->popup_centered();
-			insert_query = true;
+	if (bool(EDITOR_DEF("editors/animation/confirm_insert_track", true)) && num_tracks > 0) {
+		//potential new key, does not exist
+		if (num_tracks == 1) {
+			// TRANSLATORS: %s will be replaced by a phrase describing the target of track.
+			insert_confirm_text->set_text(vformat(TTR("Create new track for %s and insert key?"), last_track_query));
 		} else {
-			call_deferred(SNAME("_insert_delay"), reset_allowed && EDITOR_GET("editors/animation/default_create_reset_tracks"), all_bezier && EDITOR_GET("editors/animation/default_create_bezier_tracks"));
-			insert_queue = true;
+			insert_confirm_text->set_text(vformat(TTR("Create %d new tracks and insert keys?"), num_tracks));
 		}
 
+		insert_confirm_bezier->set_visible(all_bezier);
+		insert_confirm_reset->set_visible(reset_allowed);
+
+		insert_confirm->get_ok_button()->set_text(TTR("Create"));
+		insert_confirm->popup_centered();
 	} else {
-		if (!insert_query && !insert_queue) {
-			// Create Beziers wouldn't make sense in this case, where no tracks are being created
-			call_deferred(SNAME("_insert_delay"), reset_allowed && EDITOR_GET("editors/animation/default_create_reset_tracks"), false);
-			insert_queue = true;
-		}
+		_insert_track(reset_allowed && EDITOR_GET("editors/animation/default_create_reset_tracks"), all_bezier && EDITOR_GET("editors/animation/default_create_bezier_tracks"));
 	}
+
+	insert_queue = false;
 }
 
-void AnimationTrackEditor::_insert_delay(bool p_create_reset, bool p_create_beziers) {
-	if (insert_query) {
-		//discard since it's entered into query mode
-		insert_queue = false;
-		return;
+void AnimationTrackEditor::_query_insert(const InsertData &p_id) {
+	if (!insert_queue) {
+		insert_data.clear();
 	}
 
+	for (const InsertData &E : insert_data) {
+		//prevent insertion of multiple tracks
+		if (E.path == p_id.path) {
+			return; //already inserted a track for this on this frame
```

Comment:
While at it, could improve the comment style:
```suggestion
		// Prevent insertion of multiple tracks.
		if (E.path == p_id.path) {
			return; // Already inserted a track for this on this frame.
```